client/src/unifycr-sysio.c

```diff
@@ -158,41 +169,49 @@ int UNIFYCR_WRAP(rename)(const char *oldpath, const char *newpath)
     /* check whether the old path is in our file system */
     if (unifycr_intercept_path(oldpath)) {
         /* for now, we can only rename within our file system */
-        if (! unifycr_intercept_path(newpath)) {
+        if (!unifycr_intercept_path(newpath)) {
             /* ERROR: can't yet rename across file systems */
             errno = EXDEV;
             return -1;
         }
 
         /* verify that we really have a file by the old name */
         int fid = unifycr_get_fid_from_path(oldpath);
-        DEBUG("orig file in position %d\n", fid);
         if (fid < 0) {
             /* ERROR: oldname does not exist */
             DEBUG("Couldn't find entry for %s in UNIFYCR\n", oldpath);
             errno = ENOENT;
             return -1;
         }
+        DEBUG("orig file in position %d\n", fid);
+
+        /* check that new name is within bounds */
+        size_t newpathlen = strlen(newpath) + 1;
+        if (newpathlen > UNIFYCR_MAX_FILENAME) {
+            errno = ENAMETOOLONG;
+            return -1;
+        }
+
+        /* TODO: rename should replace existing file atomically */
 
         /* verify that we don't already have a file by the new name */
-        if (unifycr_get_fid_from_path(newpath) < 0) {
-            /* check that new name is within bounds */
-            size_t newpathlen = strlen(newpath) + 1;
-            if (newpathlen > UNIFYCR_MAX_FILENAME) {
-                errno = ENAMETOOLONG;
+        int newfid = unifycr_get_fid_from_path(newpath);
+        if (newfid >= 0) {
+            /* something exists in newpath, need to delete it */
+            int ret = UNIFYCR_WRAP(unlink)(newpath);
```

Comment:
Not sure if it's an issue, but the compiler is issuing a warning on this:
```
../../../client/src/unifycr-sysio.c:201:13: warning: implicit declaration of function ‘__wrap_unlink’ [-Wimplicit-function-declaration]
             int ret = UNIFYCR_WRAP(unlink)(newpath);
             ^
```

Reply:
Nice catch.  That looks worrisome.  I'll dig into that.

Reply:
Found the problem with UNIFYCR_WRAP here, will fix that in a separate PR.